src/index.js

```diff
@@ -7,50 +7,95 @@ module.exports = {
       // default config
       {
         batchSize: 1,
-        insertOptions: { w: 1 },
+        operationType: 'insert'
       },
-      // overrided options
-      options,
+      // override config
+      options
     );
 
-    // those variables can't be initialized without Promises, so we wait first drain
+    // These variables can't be initialized without Promises, so we wait for the first drain
     let dbConnection;
     let collection;
     let records = [];
 
-    // this function is usefull to insert records and reset the records array
-    const insert = async () => {
-      await collection.insert(records, config.insertOptions);
-      records = [];
+    // Supported operations
+    const operations = {
+      insert: async () => {
+        return collection.insertMany(records);
+      },
+      update: async () => {
+        // make sure we have some index so we can select the correct record to update
+        if (typeof config.indexName === 'undefined') {
+          return Promise.reject(Error('Operation type was update, but no index was provided.'));
+        }
+
+        // update each record in tandem.
+        return Promise.all(records.map(doc =>
+          collection.updateMany({ [config.indexName]: doc[config.indexName] }, { $set: doc })));
+      },
+      delete: async () => {
+        // make sure we have some index so we can select the correct record to update
+        if (typeof config.indexName === 'undefined') {
+          return Promise.reject(Error('Operation type was delete, but no index was provided.'));
+        }
+
+        // update each record in tandem.
+        return Promise.all(records.map(doc =>
+          collection.deleteMany({ [config.indexName]: doc[config.indexName] }, { $set: doc })));
+      },
+      invalid: () => { return Promise.reject(Error(`Invalid operation type: ${config.operationType}`)); }
+    };
+
+    // Utility for writing to the db with the correct operation
+    const writeToDB = async () => {
```

Comment:
I think we could cache this function result before calling it.
Here you call this function every time we need to write to db (line 84).

Reply:
Plus the error about invalid operation will be triggered before the stream is used.

Reply:
I speak about which operation to choose (which callback)